crates/biome_analyze/src/profiling.rs

```diff
@@ -0,0 +1,477 @@
+    fn new_enabled(label: RuleLabel) -> Self {
+        Self {
+            label: Some(label),
+            #[cfg(not(target_arch = "wasm32"))]
+            start: Instant::now(),
+        }
+    }
+
+    fn new_disabled() -> Self {
+        // We still initialize `start` to a valid Instant to keep struct layout simple,
+        // but it won't be used as `label` is None.
+        Self {
+            label: None,
+            #[cfg(not(target_arch = "wasm32"))]
+            start: Instant::now(),
+        }
+    }
+
+    /// Consume the timer and manually record the elapsed time (useful when you need explicit control).
+    pub fn stop(self) {
+        drop(self)
+    }
+}
+
+impl Drop for RuleRunTimer {
+    fn drop(&mut self) {
+        // We use Drop to record the elapsed time so its impossible to accidentally reuse the timer.
+        if let Some(label) = self.label.take() {
+            #[cfg(not(target_arch = "wasm32"))]
+            let elapsed = self.start.elapsed();
+            #[cfg(target_arch = "wasm32")]
+            let elapsed = Duration::ZERO;
+            with_profiler(|p| p.record(label, elapsed));
+        }
+    }
+}
+
+/// Starts measuring execution time for a built-in rule `R`.
+///
+/// When profiling is disabled, returns a no-op timer with near-zero overhead.
+pub fn start_rule<R: Rule>() -> RuleRunTimer {
+    if !is_enabled() {
+        return RuleRunTimer::new_disabled();
+    }
+    let key: RuleKey = RuleKey::rule::<R>();
+    RuleRunTimer::new_enabled(RuleLabel::builtin(key.group(), key.rule_name()))
+}
+
+/// Starts measuring execution time for a plugin rule with the specified `name`.
+///
+/// When profiling is disabled, returns a no-op timer with near-zero overhead.
+pub fn start_plugin_rule(name: impl Into<Box<str>>) -> RuleRunTimer {
+    if !is_enabled() {
+        return RuleRunTimer::new_disabled();
+    }
+    RuleRunTimer::new_enabled(RuleLabel::plugin(name))
+}
+
+/// Records a duration for the given label, bypassing RAII timers.
+///
+/// Useful for one-off custom measurements.
+pub fn record_rule_time(label: RuleLabel, delta: Duration) {
+    if !is_enabled() {
+        return;
+    }
+    with_profiler(|p| p.record(label, delta));
+}
+
+/// Returns a snapshot of all collected profiles in unspecified order.
+pub fn snapshot() -> Vec<RuleProfile> {
+    with_profiler(|p| p.snapshot()).unwrap_or_default()
+}
+
+/// Returns all profiles sorted by total time (descending).
+pub fn drain_sorted_by_total(reset_after: bool) -> Vec<RuleProfile> {
+    let mut profiles = with_profiler(|p| p.snapshot()).unwrap_or_default();
+
+    profiles.sort_by(|a, b| b.total.cmp(&a.total));
+
+    if reset_after {
+        reset();
+    }
+
+    profiles
+}
+
+/// Clears all collected metrics.
+pub fn reset() {
+    with_profiler(|p| p.reset());
+}
+
+/// Utility for formatting a summary for display purposes.
+pub fn format_profiles_table(mut profiles: Vec<RuleProfile>, top_n: Option<usize>) -> MarkupBuf {
```

Comment:
You should create a type e.g. `struct DisplayProfiles` and then implement `biome_console::fmt::Display`. 

Then you can do `markup!{{ DisplayProfiles(profiles, top_n) }}`.

Doing so allows better writing in the buffer, where we can write per line, which we can't do with `markup!`. Plus, we don't need to allocate any `MarkupBuf`